rfc/bh-rfc-2.md

```diff
@@ -0,0 +1,114 @@
+---
+bh-rfc: 2
+title: Early Return Pattern
+authors: |
+    [Holms, Alyx](aholms@specterops.io)
+    [Pomeroy, Kaleb](kpomeroy@specterops.io)
+status: DRAFT
+created: 2025-04-15
+---
+
+# Early Return Pattern
+
+Go encourages keeping blocks small, variables scoped tightly, and failure paths easy to follow. While Inline Initializer Statements help with scoping, they should not be combined with chained nesting patterns, which obscure control flow and increase complexity. Nesting hides the success case in the deepest block, when that’s actually the flow we care about most.
+
+## 1. Overview
+
+This proposal recommends adopting the Early Return Pattern as the default style for handling sequential workflows and error propagation in Go. Early returns flatten control flow, make error handling explicit, and align with Go’s idiomatic practices as described in Effective Go. This change improves readability, reduces indentation depth, and creates more maintainable and composable functions.
+
+### 1.1 Example of Early Return Pattern
+
+```
+func flat() {
+	a, err := stepA()
+	if err != nil {
+		fmt.Println("A had an error")
+		return
+	}
+
+	b, err := stepB()
+	if err != nil {
+		fmt.Println("B had an error")
+		return
+	}
+	c, err := stepC()
+	if err != nil {
+		fmt.Println("C had an error")
+		return
+	}
+
+	fmt.Println(a + b + c)
+}
+```
+
+### 1.2 Example of an Early Return Anti-pattern
+
+```
+func nesting() {
+	if a, err := stepA(); err != nil {
+		fmt.Println("A had an error")
+	} else if b, err := stepB(); err != nil {
+		fmt.Println("B had an error")
+	} else if c, err := stepC(); err != nil {
+		fmt.Println("C had an error")
+	} else {
+		fmt.Println(a + b + c)
+	}
+}
+```
+
+## 2. Motivation & Goals
+
+The goal of this RFC is to establish a consistent and idiomatic approach to control flow and error handling in sequential Go functions. Specifically, it promotes the use of early returns over chained nested if statements when dealing with dependent operations that may return errors.
+
+-   **Readability** Early returns keep the success path at the top level, making it easier to understand the primary logic of a function at a glance.
+-   **Maintainability** Flattened control flow reduces indentation and visual noise, making future changes and code reviews easier.
+-   **Idiomatic Go** This approach aligns with community conventions and guidance from Effective Go and widely adopted Go style guides.
+-   **Error clarity** By separating failure paths early, each step’s error handling becomes clearer and less entangled with the main logic.
+-   **Reduced coupling**: Avoids tightly chaining variable scopes, enabling greater reuse and flexibility of intermediate values.
```

Comment:
I know the kind of refactors this would be helpful for, but messing with intermediates can just as easily become spaghetti as what we have now if not guarded against abuse. How will we allow for this reduced coupling while promoting a style guide that helps curb poor use of access to variables? Specifically, how do we help shape code such that you either build control flow or business logic in a function, but never both at the same time?